torch/nested/_internal/ops.py

```diff
@@ -1114,12 +1114,90 @@ def native_layer_norm_default(func, *args, **kwargs):
     )
 
     inp = new_kwargs.pop("input")
-    normalized_shape = new_kwargs["normalized_shape"]
 
-    # Ensure we're not trying to normalize over the ragged dim
-    if inp.dim() < 3 or (inp.dim() - len(normalized_shape)) < 2:
+    if inp.dim() <= 2:
         raise RuntimeError(
-            "layer_norm(): normalizing over ragged dim not supported for nested tensors"
+            "layer_norm(): not supported for NestedTensor objects with 2 or fewer dimensions"
+        )
+
+    normalized_shape = new_kwargs[
+        "normalized_shape"
+    ]  # denote -1 as the ragged dimension, i.e. normalize on the ragged dimension
+
+    num_output_dims = inp.dim() - len(normalized_shape)
+
+    if num_output_dims == 0:  # error if trying to normalize over the batch dimension
+        raise RuntimeError(
+            "layer_norm(): not supported when normalizing over the batch dimension for NestedTensor"
+        )
+
+    if -1 in normalized_shape and inp._lengths is not None:
+        raise RuntimeError(
+            "layer_norm(): not supported where lengths is not None if operating on the ragged dimension for NestedTensor"
+        )
+
+    if (
+        -1 in normalized_shape
+    ):  # special handling for normalizing over the ragged dimension
```

Comment:
Could you change the handling here?

-1 makes sense, but we could also get a nested int here, e.g. j0 or something.

I feel like it might be better to say "if len(normalized_shape) == len(self._size) - 1` (not sure if that's the right way to get the size).

Also, I think technically there's another case for `len(normalized_shape) == len(self._size)`, which isn't handled here. I think it's fine to not handle it but maybe add an error.

Reply:
> -1 makes sense, but we could also get a nested int here, e.g. j0 or something.

i was mixing up NSTs (which don't support shape on the ragged dimension) with NJTs (which do support it); thanks for pointing this out!

> I feel like it might be better to say "if len(normalized_shape) == len(self._size) - 1` (not sure if that's the right way to get the size).

i think this would need to be changed in the future when we handle `jt_all` (`ragged_idx > 1`). for example, if (in the future) we try to normalize a nt of shape `(4, 5, j0, 6, 7)` on `(5, j0, 6, 7)`, checking the lengths would become inaccurate!

> there's another case for len(normalized_shape) == len(self._size)

would this be the same as checking `inp.dim() - len(normalized_shape) == 0` (aka normalizing on batch dimension)? if so, i think this is handled and throws an error!

also, changed `num_output_dims` to a variable name that's more descriptive :)